pf2/tensor.py

```diff
@@ -2,55 +2,54 @@
 import numpy as np
 from anndata import AnnData
 from pacmap import PaCMAP
-from parafac2 import parafac2_nd
-from tlviz.factor_tools import degeneracy_score
+from parafac2.parafac2 import parafac2_nd, store_pf2
+import scipy.cluster.hierarchy as sch
+from scipy.stats import gmean
+from sklearn.linear_model import LinearRegression
 
-OPTIMAL_RANK = 43
+OPTIMAL_RANK = 40
 
 
-def store_pf2(
-    data: AnnData, parafac2_output: tuple[np.ndarray, list, list]
-) -> AnnData:
-    """Store the Pf2 results into the anndata object."""
-    sg_index = data.obs["condition_unique_idxs"]
-
-    data.uns["Pf2_weights"] = parafac2_output[0]
-    data.uns["Pf2_A"], data.uns["Pf2_B"], data.varm["Pf2_C"] = parafac2_output[
-        1
-    ]
-    data.uns["Pf2_rank"] = data.uns["Pf2_A"].shape[1]
-
-    data.obsm["projections"] = np.zeros(
-        (data.shape[0], len(data.uns["Pf2_weights"]))
-    )
-    for i, p in enumerate(parafac2_output[2]):
-        data.obsm["projections"][sg_index == i, :] = p  # type: ignore
-
-    data.obsm["weighted_projections"] = (
-        data.obsm["projections"] @ data.uns["Pf2_B"]
-    )
-
-    return data
-
 
 def pf2(
     data: AnnData,
     rank: int = OPTIMAL_RANK,
     random_state=1,
     do_embedding: bool = True,
 ) -> anndata.AnnData:
-    pf_out, r2x = parafac2_nd(data, rank=rank, random_state=random_state)
+    pf_out, r2x = parafac2_nd(data, rank=rank, random_state=random_state, tol=1e-7)
 
     data = store_pf2(data, pf_out)
 
-    print(f"Degeneracy score: {degeneracy_score((pf_out[0], pf_out[1]))}")
-
     if do_embedding:
         pcm = PaCMAP(random_state=random_state)
-        data.obsm["embedding"] = pcm.fit_transform(data.obsm["projections"])  # type: ignore
-        pcm = PaCMAP(random_state=random_state)
-        data.varm["embedding"] = pcm.fit_transform(data.varm["Pf2_C"])  # type: ignore
-        pcm = PaCMAP(random_state=random_state)
-        data.uns["embedding"] = pcm.fit_transform(data.uns["Pf2_A"])  # type: ignore
+        data.obsm["X_pf2_PaCMAP"] = pcm.fit_transform(data.obsm["projections"])  # type: ignore
 
     return data, r2x
+
+
+def correct_conditions(X: anndata.AnnData):
+    """Correct the conditions factors by overall read depth."""
+    sgIndex = X.obs["condition_unique_idxs"]
+    counts = np.zeros((np.amax(sgIndex) + 1, 1))
+
+    cond_mean = gmean(X.uns["Pf2_A"], axis=1)
```

Comment:
@JacksonLChin @aarmey The error starts here in the geometric mean, even though the values are not 0 (although many of the condition factors are 1e-4 or less) there is an error on the `ln`. This hasn't been an issue before which I find puzzling, but the gmean some some of the components with very small values becomes NaN

Reply:
I've encountered this error with the dataset before--I've found that some patients have small, negative component values that throw-off the geometric mean. I've swapped out geometric for arithmetic mean before, though I'm not certain how that'll influence the factors. Might also be useful to discard patients with very low cell counts--I'm guessing this might be related to patients with few/none of certain cell types.

Reply:
Yeah there is definitely a lot of variation and patients with very small cell counts. 
<img width="1025" alt="Screenshot 2024-06-06 at 12 29 11 PM" src="https://github.com/meyer-lab/BAL-Pf2/assets/91349614/d5df6bf5-b8e4-4224-9523-6f65b5093d87">

Reply:
I think it's ok to you the arithmetic mean here for now. Other constraints on the Pf2 fitting may fix this issue in the future.

Reply:
Sounds good! Should removing patient samples with a small cell count happen before or after factorization? (I think before might make more sense in our case)

Reply:
Before is probably better. There's not much we can do with them after all.